pixels-index/pixels-index-rocksdb/src/main/java/io/pixelsdb/pixels/index/rocksdb/RocksDBIndex.java

```diff
@@ -398,32 +439,35 @@ public boolean closeAndRemove() throws SinglePointIndexException
     // Convert IndexKey to byte array
     private static byte[] toByteArray(IndexProto.IndexKey key)
     {
-        byte[] indexIdBytes = ByteBuffer.allocate(Long.BYTES).putLong(key.getIndexId()).array(); // Get indexId bytes
-        byte[] keyBytes = key.getKey().toByteArray(); // Get key bytes
-        byte[] timestampBytes = ByteBuffer.allocate(Long.BYTES).putLong(key.getTimestamp()).array(); // Get timestamp bytes
+        byte[] indexIdBytes = ByteBuffer.allocate(Long.BYTES).putLong(key.getIndexId()).array();
+        byte[] keyBytes = key.getKey().toByteArray();
+        byte[] timestampBytes = ByteBuffer.allocate(Long.BYTES).order(ByteOrder.BIG_ENDIAN).putLong(key.getTimestamp()).array();
         // Combine indexId, key and timestamp
-        byte[] compositeKey = new byte[indexIdBytes.length + 1 + keyBytes.length + 1 + timestampBytes.length];
+        byte[] compositeKey = new byte[indexIdBytes.length + keyBytes.length + timestampBytes.length];
         // Copy indexId
         System.arraycopy(indexIdBytes, 0, compositeKey, 0, indexIdBytes.length);
-        // Add separator
-        compositeKey[indexIdBytes.length] = ':';
         // Copy key
-        System.arraycopy(keyBytes, 0, compositeKey, indexIdBytes.length + 1, keyBytes.length);
-        // Add separator
-        compositeKey[indexIdBytes.length + 1 + keyBytes.length] = ':';
+        System.arraycopy(keyBytes, 0, compositeKey, indexIdBytes.length, keyBytes.length);
         // Copy timestamp
-        System.arraycopy(timestampBytes, 0, compositeKey, indexIdBytes.length + 1 + keyBytes.length + 1, timestampBytes.length);
+        System.arraycopy(timestampBytes, 0, compositeKey, indexIdBytes.length + keyBytes.length, timestampBytes.length);
 
         return compositeKey;
     }
 
+    // Extract key prefix (key without timestamp) from key
+    private static byte[] toKeyPrefix(IndexProto.IndexKey key)
+    {
+        byte[] fullKey = toByteArray(key);
+        int prefixLength = fullKey.length - Long.BYTES;
+        return Arrays.copyOf(fullKey, prefixLength);
+    }
+
     // Create composite key with rowId
     private static byte[] toNonUniqueKey(byte[] keyBytes, byte[] valueBytes)
     {
-        byte[] nonUniqueKey = new byte[keyBytes.length + 1 + valueBytes.length];
+        byte[] nonUniqueKey = new byte[keyBytes.length + valueBytes.length];
```

Comment:
Compositing a non-unique key like this causes memory and CPU overhead. Build it from scratch.